File: tutorials/src/main/java/com/android/commons/base/BaseActivity.java
package com.android.commons.base;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewbinding.ViewBinding;

public abstract class BaseActivity<VB extends ViewBinding, VM extends BaseViewModel>
        extends AppCompatActivity implements BaseViews {

    public VB binding;
    public VM viewModel;

    public abstract VB createViewBinding();
    public abstract VM createViewModel();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = createViewModel();
        binding = createViewBinding();

        initializeViews();
        initializeEvents();
    }
}

File: vnCalling/app/src/main/java/com/android/vncalling/base/BaseActivity.kt
package com.android.vncalling.base

import android.os.Bundle
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.viewbinding.ViewBinding

abstract class BaseActivity<VB : ViewBinding, VM : BaseViewModel> : AppCompatActivity() {

    lateinit var viewModel: VM
    lateinit var binding: VB

    abstract fun createViewModel(): VM
    abstract fun getViewBinding(): VB
    abstract fun initialize()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = createViewModel()

        this.binding = getViewBinding()
        setContentView(this.binding.root)

        initialize()
    }

    fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    fun showToast(@StringRes message: Int) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }
}